Adds facebox

robmarkcole opened the pull request:
Replaces https://github.com/home-assistant/home-assistant/pull/14279

## Description:
Adds component for face detection (number of faces) and identification (recognition of taught faces) using [facebox](https://machineboxio.com/docs/facebox/teaching-facebox). Run facebox with:
```
MB_KEY="INSERT-YOUR-KEY-HERE"

docker run -p 8080:8080 -e "MB_KEY=$MB_KEY" machinebox/facebox
```

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io/pull/5300) with documentation (if applicable):** home-assistant/home-assistant.github.io#<5300>

## Example entry for `configuration.yaml` (if applicable):
```yaml
image_processing:
  - platform: facebox_face_detect
    ip_address: 192.168.0.1
    port: 8080
    source:
      - entity_id: camera.local_file
```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

robmarkcole committed:
Adds facebox-detect

robmarkcole commented:
@pvizeli will keep this initial release a MVP

MartinHjelmare commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,90 @@
+https://home-assistant.io/components/image_processing.facebox_face_detect
+"""
+import base64
+import requests
```
This is a 3rd party library and should be grouped with those ie voluptuous.

robmarkcole replied:
Ok

MartinHjelmare commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,90 @@
+            response = requests.post(
+                self._url,
+                json=self.encode_image(image),
+                timeout=30
```
Do we need a 30 second timeout? I suggest 9 seconds, to avoid slow update warning.

robmarkcole replied:
Ok

MartinHjelmare commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,90 @@
+            self.faces = response['faces']
+
+        else:
+            self.total_faces = "Request_failed"
```
This would change the type of total faces from integer to string. That will make it hard to use templates for this attribute.

I suggest that we don't update the total faces if there's an error. Or set it to `0`.

robmarkcole replied:
Or None? I don’t want to have users think a classification has been successful and that the face count is zero, when in face the classification was unsuccessful

MartinHjelmare replied:
`None` is good.

MartinHjelmare commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,90 @@
+            response['success'] = False
+
+        if response['success']:
+            self.total_faces = response['facesCount']
```
We shouldn't update these instance attributes directly but pass the faces and faces count through the `process_faces` method. That will fire an event with the face data and update the instance attributes.

robmarkcole replied:
Face detect and face identify components work differently (confusingly). Face detect (this Pr) have a state which is the number of faces, whilst face identify components have a state which is a recognised face. I prefer the former so have avoided using process face. If we want this to be a face identify component I will use process_faces

MartinHjelmare replied:
We can still use `process_faces`. It will not fire events if there's no confidence set and no confidence attribute in the face dict. Just make sure `faces` and `total` are of correct types.

MartinHjelmare commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,90 @@
+        else:
+            self._name = "Facebox {0}".format(
+                split_entity_id(camera_entity)[1])
+        self.total_faces = 0
```
This is already set in the parent class.

robmarkcole replied:
Good

MartinHjelmare commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,90 @@
+            self._name = "Facebox {0}".format(
+                split_entity_id(camera_entity)[1])
+        self.total_faces = 0
+        self.faces = []
```
Same as above.

robmarkcole replied:
Ok

MartinHjelmare commented on tests/components/image_processing/test_facebox_face_detect.py:
```diff
@@ -0,0 +1,68 @@
+    }
+
+
+class TestFaceboxSetup(object):
```
You don't need to use a class for the tests. Just define test functions that have names starting with `test_` Pass in `hass` as parameter to the functions to get a hass instance to test with. Setup and teardown is handled by pytest.

robmarkcole replied:
Ok

MartinHjelmare commented on tests/components/image_processing/test_facebox_face_detect.py:
```diff
@@ -0,0 +1,68 @@
+
+        with requests_mock.Mocker() as mock_req:
+            url = "http://{}:{}/facebox/check".format(MOCK_IP, MOCK_PORT)
+            mock_req.get(url, text=MOCK_RESPONSE)
```
Use the `post` method on the mock object to mock a post request.

https://requests-mock.readthedocs.io/en/latest/mocker.html#methods

robmarkcole commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,90 @@
+import base64
+import requests
+import logging
+import voluptuous as vol
```
Ok

robmarkcole commented on homeassistant/components/image_processing/facebox_face_detect.py:
```diff
@@ -0,0 +1,90 @@
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_IP_ADDRESS): cv.string,
+    vol.Required(CONF_PORT): cv.string,
```
Ok

robmarkcole committed:
Drop face_detect 

drop face_detect from filename and component contents as this component will handle both detection and identification

robmarkcole committed:
Update and rename test_facebox_face_detect.py to test_facebox.py

robmarkcole committed:
Adds matched_faces attr

robmarkcole committed:
Update test_facebox.py

MartinHjelmare commented on homeassistant/components/image_processing/facebox.py:
```diff
@@ -0,0 +1,109 @@
+        base64_img = base64.b64encode(image).decode('ascii')
+        return {"base64": base64_img}
+
+    def get_matched_faces(self, response):
```
Maybe pass in `faces` as that's what we're using in the method?

MartinHjelmare reviewed (commented):
I said something inaccurate in a previous comment about `process_faces` and when it will fire events. Actually it's like this: it will always fire a face detected event if `self.confidence` is not truthy.

I still think we should use `process_faces`. It should give us the proper behavior for this implementation.

robmarkcole committed:
Amend get_matched_faces to take faces

robmarkcole commented:
@MartinHjelmare I agree that `process_faces` should be used for consistency. I think it's just the test to fix now? Btw re test using `classes` or `def`, appears most legacy tests use `classes`, but now the best practice is to use `def`? Would be good to document this kind of best practice in the docs

robmarkcole commented:
@MartinHjelmare I've basically copied the tests from the [microsoft component](https://github.com/home-assistant/home-assistant/blob/03fb062b3fdab44eea797a37225aa4675f240bf4/tests/components/image_processing/test_microsoft_face_detect.py) but am getting errors which appear to relate to asyncio - any ideas what's wrong here?
```
tests/components/image_processing/test_facebox.py:43:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
homeassistant/setup.py:30: in setup_component
    async_setup_component(hass, domain, config), loop=hass.loop).result()
../../../miniconda3/lib/python3.6/concurrent/futures/_base.py:427: in result
    self._condition.wait(timeout)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Condition(<unlocked _thread.RLock object owner=0 count=0 at 0x10f8536f0>, 0)>, timeout = None
```

MartinHjelmare commented on tests/components/image_processing/test_facebox.py:
```diff
@@ -0,0 +1,62 @@
+    }
+
+
+def test_setup_platform(hass):
```
Make the tests coroutines with `async def`.

MartinHjelmare reviewed (commented):
Yeah I'm not sure exactly why it fails with sync tests, but coroutines should work.

We should probably look into this some time. It's useful to not have to use coroutines in the tests when setting up components too.

robmarkcole committed:
Update test_facebox.py

robmarkcole commented:
@MartinHjelmare have swapped in `async_setup_component` but setup is failing:
```
AssertionError: setup_component failed, expected 1 got 0: None
```

robmarkcole committed:
Update test_facebox.py

robmarkcole committed:
Update test_facebox.py

robmarkcole committed:
Make mock_req.post

was mock_req.get

MartinHjelmare commented on tests/components/image_processing/test_facebox.py:
```diff
@@ -0,0 +1,57 @@
+    with requests_mock.Mocker() as mock_req:
+        url = "http://{}:{}/facebox/check".format(MOCK_IP, MOCK_PORT)
+        mock_req.post(url, text=MOCK_RESPONSE)
+        ip.scan(hass, entity_id=VALID_ENTITY_ID)
```
Call the service directly with `await hass.services.async_call(...)`. Since we're in a coroutine here, we have to use the hass async api consistently.

robmarkcole replied:
ok

MartinHjelmare commented on tests/components/image_processing/test_facebox.py:
```diff
@@ -0,0 +1,57 @@
+        url = "http://{}:{}/facebox/check".format(MOCK_IP, MOCK_PORT)
+        mock_req.post(url, text=MOCK_RESPONSE)
+        ip.scan(hass, entity_id=VALID_ENTITY_ID)
+        hass.block_till_done()
```
`await hass.async_block_till_done()`

robmarkcole replied:
ok

robmarkcole committed:
Update test_facebox.py

robmarkcole commented on tests/components/image_processing/test_facebox.py:
```diff
@@ -0,0 +1,59 @@
+        mock_req.post(url, text=MOCK_RESPONSE)
+        await hass.services.async_call(ip.DOMAIN,
+                                       'scan',
+                                       {'entity_id': VALID_ENTITY_ID})
```
@MartinHjelmare this is giving confusing error:
```
simplejson.errors.JSONDecodeError: Expecting value: line 3 column 12 (char 30)
core.py                    399 INFO     Bus:Handling <Event service_executed[L]: service_call_id=4368981800-1>
```

robmarkcole committed:
Update test_facebox.py

robmarkcole commented:
@MartinHjelmare any ideas about the test service call which is giving the error:
```
entity.py                  206 ERROR    Update for image_processing.facebox_demo_camera fails
Traceback (most recent call last):
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/helpers/entity.py", line 204, in async_update_ha_state
    yield from self.async_device_update()
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/helpers/entity.py", line 325, in async_device_update
    yield from self.async_update()
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/components/image_processing/__init__.py", line 146, in async_update
    yield from self.async_process_image(image.content)
  File "/Users/robincole/miniconda3/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/components/image_processing/facebox.py", line 67, in process_image
    timeout=9
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/requests/models.py", line 892, in json
    return complexjson.loads(self.text, **kwargs)
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/simplejson/__init__.py", line 518, in loads
    return _default_decoder.decode(s)
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 3 column 12 (char 30)
core.py                    399 INFO     Bus:Handling <Event service_executed[L]: service_call_id=4357840456-1>
```

MartinHjelmare commented:
Rebase on latest dev. There was a problem with the clean up PR of custom components loading, so Paulus reverted that. Should work as before now. I think that's why your test failed, cause you're using a custom component in the test.

robmarkcole committed:
Rename to facebox

Multiple changes to address reviewer comments

robmarkcole committed:
Merge remote-tracking branch 'home-assistant/dev' into facebox-detect

robmarkcole committed:
Merge remote-tracking branch 'origin/facebox-detect' into facebox-detect

robmarkcole commented:
@MartinHjelmare I've rebased on latest but still same error from the test unfortunately

robmarkcole committed:
Working test

robmarkcole committed:
Update test with matched_faces

This test should fail

robmarkcole commented:
@MartinHjelmare I fixed that error - it was not due to the custom component but required a `patch`. 
I now have another error, the test should fail but it passes and I get:
```
RuntimeWarning: coroutine 'test_process_image' was never awaited
```

MartinHjelmare commented:
The call to `encode_image` will take place and actually read the image from the demo camera. I think we need to mock an image that will properly pass that method and find out why the demo camera images didn't work.

Have you tested the method for real with real images? You can see if you get the same error when you run the method directly with one of the images from the demo camera. To make sure we can reproduce and pinpoint the error.

Probably mock `...camera.async_get_image` and set the return value to an image content that works for the test. I'm not sure of exactly the correct path to mock.

mikedm139 committed:
Update hitron_coda.py to fix login for Shaw modems (#14306)

I have a Hitron modem provided by Shaw communications rather than from Rogers as the Docs specify for this device_tracker but it seems like the api/code is all the same except that the login failed due to the password being passed as "pws" instead of "pwd". Making that one character change allowed HASS to read the connected device details from my Hitron modem. If this difference is actually one that stands between the Rogers-provided Hitron modems and the Shaw-provided variant, I am happy to create another device-tracker file for the Shaw modem. I just figured I would go with the simplest solution first.

gerard33 committed:
Add sensors for BMW electric cars (#14293)

* Add sensors for electric cars

* Updates based on review of @MartinHjelmare

* Fix Travis error

* Another fix for Travis

worm-ee committed:
Add more homematicip cloud components (#14084)

* Add support for shutter contact and motion detector device

* Add support for power switch devices

* Add support for light switch device

* Cleanup binary_switch and light platform

* Update comment

MartinHjelmare commented:
Patch decorator might not be async friendly. Try patching inside the test function instead.

m4dmin committed:
add 2 devices (#14321)

* add 2 devices

io:RollerShutterUnoIOComponent
io:ExteriorVenetianBlindIOComponent

* add 2 devices

* Update tahoma.py

* Fix hounci-bot violation

* Fixed Travis CI build failure

./homeassistant/components/cover/tahoma.py:83:13: E125 continuation line with same indent as next logical line

* Fixed Travis CI build failure

E125 continuation line with same indent as next logical line

* Fixed Travis CI build failure

E127 continuation line over-indented for visual indent

* Fix indent

* Change check

dlbroadfoot committed:
Gogogate2 0.1.1 (#14294)

* Gogogate2 - bump version

Uses latest version of library which ensures commands to device are idempotent

* Update requirements_all

* Expose sensor temperature

* update version

* import attribute

* Set temperature

* Remove temperature attribute

Removed temperature attribute until it can be re-implemented as a separate sensor.

* Update ordering

* Fix copy-&-paste issue

todschmidt committed:
Snips: Added slot values for siteId and probability (#14315)

* Added solt values for siteId and probability

* Update snips.py

* Update test_snips.py

balloob committed:
Fix Insteon PLM coverage

stephanerosi committed:
Add help for conversation/process service (#14323)

* Add help for conversation/process service

* Add logging to debug text received when service is called

* Move conversation to specific folder

nickw444 committed:
Fix BOM weather '-' value (#14042)

robmarkcole commented:
@MartinHjelmare moved patch (not sure about indentation) but still same error:
```
    with requests_mock.Mocker() as mock_req:
        url = "http://{}:{}/facebox/check".format(MOCK_IP, MOCK_PORT)
        mock_req.post(url, text=MOCK_RESPONSE)
        data = {ATTR_ENTITY_ID: VALID_ENTITY_ID}
        with patch('homeassistant.components.camera.demo.DemoCamera.camera_image', autospec=True, return_value=b'Test'):
            await hass.services.async_call(ip.DOMAIN,
                                           ip.SERVICE_SCAN,
                                           service_data=data)
```

andrey-git committed:
Make sure zwave nodes/entities enter the registry is proper state. (#14251)

* When zwave node's info is parsed remove it and re-add back.

* Delay value entity if not ready

* If node is ready consider it parsed even if manufacturer/product are missing.

* Add annotations

MartinHjelmare commented:
Ok, I managed to fix the test. I was mistaken earlier that the call to `encode_image` was the problem. It was just the mock request didn't handle the data as json but as regular text.

See my commit here:
https://github.com/MartinHjelmare/home-assistant/commit/2e0435ecaec3b3db5930cfcd33584cac54d19817

I can push this to your branch if you like or if you want to fix it yourself. Just let me know.

evgeniy-khatko committed:
Improving icloud device tracker (#14078)

* Improving icloud device tracker

* Adding config validations for new values

* Adding config validations for new values

* Moving icloud specific setup to platform schema. Setting default in platform schema.

bachya committed:
Adds useful attributes to RainMachine programs and zones (#14087)

* Starting to add attributes

* All attributes added to programs

* Basic zone attributes in place

* Added advanced properties for zones

* Working to move common logic into component + dispatcher

* We shouldn't calculate the MAC with every entity

* Small fixes

* Small adjustments

* Owner-requested changes

* Restart

* Restart part 2

* Added ID attribute to each switch

* Collaborator-requested changes

mario-tux committed:
Waze Travel Time: optional inclusive/exclusive filters (#14000)

* Waze Travel Time: optional inclusive/exclusive filters

Added optional `inc_filter` and `excl_filter' params that allow to refine the reported routes: the first is not always the best/desired. A simple case-insensitive filtering (no regular expression) is used.

* fix line lenght

* fix spaces

* Rename var

* Fix typo

* Fix missing var

balloob committed:
Bump frontend to 20180509.0

amelchio committed:
Ignore NaN values for influxdb (#14347)

* Ignore NaN values for influxdb

* Catch TypeError

snikch committed:
Add zone 3 for Onkyo media player (#14295)

* Add zone 3 for Onkyo media player

* CR Updates

* Fix travis lint errors

robmarkcole committed:
Passing test

robmarkcole committed:
Merge remote-tracking branch 'home-assistant/dev' into facebox-detect

robmarkcole committed:
Fix lints

robmarkcole committed:
Fix lints

robmarkcole commented:
@MartinHjelmare thanks for fixing the test, I wouldn't have guessed that solution. 
I think this is ready for 3rd party review now - perhaps @balloob @pvizeli @OttoWinter

MartinHjelmare commented:
The branch is whack again.

robmarkcole commented:
Oh bugger
